### ISSUE TITLE
Improve CSRF error handling on contact forms

### ISSUE THREAD

**Issue by harrislapiroff** [freedomofpress/pressfreedomtracker.us#1674]
[updated 6/15/2023]

Based on the conversation below this ticket is now to improve error handling on CSRF errors for contact forms:

- [ ] CSRF errors should show _on_ the form itself as form errors
- [ ] The form should retain user input so they can attempt to resubmit

We should also make sure we're logging these errors and ideally set up some sort of threshold alert (attn @chigby @maeve-fpf)

Whatever we do here should also be implemented on FPF and SDO after, so please open tickets for those sites when this is resolved.

***

Mig noticed at one point during testing that he got a csrf error that went away when he tried again... caching doesn't seem to be a problem but there is a csrf cookie that might be the culprit. We've only seen this once, and no STR yet... Value of the secret is changed each time a user logs in. If they are anonymous, it's per session.

Timebox task? See if we can repro.

**Comment by harrislapiroff** [comment]
Whoever works on this should spend an hour seeing if they can reproduce the issue on the prod server. Remember:

* Cloudflare cache behaves differently if you're logged in to Wagtail admin
* There are multiple forms on the site
* CSRF tokens should be unique per-session if I understand the [django docs](https://docs.djangoproject.com/en/4.2/ref/csrf/#how-it-works) correctly

**Comment by chigby** [comment]
So I was able to trigger a CSRF error on the contact form (on staging and production). The docs Harris linked say:

> For security reasons, the value of the secret is changed each time a user logs in.

So you can obtain an error by loading one of the pages with a form on it (the [contact page](https://staging.pressfreedomtracker.us/contact/) or [submit and incident page](https://staging.pressfreedomtracker.us/submit-incident/)), then logging in to the admin, then submitting the form. Is this the scenario Mig encountered? No idea, but it is something that could potentially happen unwittingly to people with admin access. Another way to intentionally trigger a CSRF failure is to delete the cookie after loading one of those pages, but doing this seems like you'd be aware your actions are going to cause a problem.

I'll also add that I don't think the CSRF tokens are stored on the Django session. They _can_ be, but only are if [`CSRF_USE_SESSIONS`](https://docs.djangoproject.com/en/3.2/ref/settings/#csrf-use-sessions) is set, which it's not on our sites. Meaning the token is stored in a cookie in our situation.

**Comment by harrislapiroff** [comment]
@mig5 @eloquence Is it possible this is what you encountered? Does this investigation seem sufficient? If you think it was something else we could maybe devote another hour to it.

**Comment by eloquence** [comment]
I'll wait for Mig to comment. PFT user data does show that he was last logged in a week ago so at least an admin login session  cookie was likely present.

**Comment by mig5** [comment]
Hi folks,

Unfortunately, the flow described is not what I encountered (I was in a dispVM at the time and I think it was unlikely that I had a session even if it dated from the same day, as I'm opening/closing dispVMs many times a day)

Furthermore, I actually just reproduced it in a completely fresh dispVM.

My original theory appears to be correct:

1. I loaded https://pressfreedomtracker.us/contact/ and https://pressfreedomtracker.us/submit-incident/ in two separate tabs on Firefox 102.12.0esr (Debian 11 dispVM in Qubes 4.1) simultaneously, in near real-time (like, I pressed enter after entering the addresses, in each tab, almost exactly at the same time - the /contact form slightly sooner than the other)

2. When I look at my developer toolbar, I can see I got issued two different csrftoken cookies in each tab. I think the 'normal' behaviour would be that it would be the same cookie if one had more gracefully navigated from one page to the other in the same tab. (I have verified this in another dispVM)

![Screenshot_2023-06-15_08-53-49](https://github.com/freedomofpress/pressfreedomtracker.us/assets/99173/f24474db-2f66-400a-ba40-ee29dbaf07a2)
![Screenshot_2023-06-15_08-54-03](https://github.com/freedomofpress/pressfreedomtracker.us/assets/99173/691da607-87bd-4aef-8340-8da3b659cac2)

3. I filled out the /contact form, successfully completed the recaptcha, and submitted the form. At this point I got the CSRF error.

![Screenshot_2023-06-15_08-54-55](https://github.com/freedomofpress/pressfreedomtracker.us/assets/99173/b721b7cc-8fd9-4eaa-9e9f-fbe3bba7b33a)

I didn't bother submitting the /submit-incident/ form but I can if you wish.

In short: loading two pages simultaneously when there is no previous cookie, sees one page 'clobber' the cookie of the other. This makes perfect sense since I think the browser can only hold one cookie of that name for that domain, across all tabs. 

I checked and confirmed that the submission of the /contact form, sent the cookie that was returned in the /submit-incident page, instead of the cookie that was issued when loading the /contact form. So because the cookie did not match the one that was expected on the /contact form (there is a 'hidden' form element that probably validates, on submission, that the token matches what is in the cookie) , one gets the error.

To be honest I think this is a corner case and a rare flow for most people and probably would be hard to fix.

**Comment by harrislapiroff** [comment]
Hm, yeah, I think I agree @mig5 this seems tricky to handle and rare as a use case when not specifically performing QA. Some amount of CSRF issues are tolerable since the worst-case is having to resubmit the form. This does make me think we should display a CSRF error inline on the form though and preserve the user's input. @chigby or @SaptakS is that something we can do, to your knowledge?

(Incidentally, I just received a CSRF error on Matomo's login screen under similar circumstances 😆)

**Comment by SaptakS** [comment]
I definitely think we should show proper error message and ask the user to resubmit. I think we can use https://docs.djangoproject.com/en/4.2/ref/settings/#std-setting-CSRF_FAILURE_VIEW and catch forbidden response to show some error like "There was some error related to CSRF token. Please re-submit the form."

**Comment by harrislapiroff** [comment]
Okay, let's make this ticket better user-facing error handling for CSRF errors. I'll update the description
